GOVERNANCE.md

```diff
@@ -8,13 +8,40 @@ This project may include the following roles. Additional roles may be adopted an
+The PyScript project implements a mixed committee/BDFL (Benevolent Dictator for Life) decision making model, where the BDFL works with the committee with the ultimate goal of advancing the project while developing its vision.
+
+**3.1.1. Rationale**
+
+_Why this model rather than any other ?_
+
+It comes down to “vision” and consistency.
+
+Given the current status and life of the PyScript project, we believe that the best way to encourage a mixed approach where the active `maintainers` group acts as a committee, effectively being at the center of the decision making process, and the BDFL act as ultimate gate keeper of the vision and the project users.
+
+**2.1. Consensus-Based Decision Making**. The project standard way of making decisions is through consensus of the Maintainers and the automatic blessing of the BDFL. While explicit agreement of all Maintainers is preferred, it is not required for consensus. Rather, the Maintainers will determine consensus based on their good faith considering a number of factors, including the dominant view of the Contributors and nature of support and objections. The Maintainers will document evidence of consensus in accordance with these requirements.
+
+In cases of controversial discussions and on exceptional cases, the BDFL has the right (and duty) to make a call to confirm or decline a specific decision. This right is to be exercised very carefully, and used only on delicate matters whenever it is fit.
+
+**2.2. Appeal Process**. Decisions may be appealed by opening an issue and that appeal will be considered by the Maintainers in good faith, who will respond in writing within a reasonable time.
+
+**2.3. BDFL Role**
+
+The main goal of the PyScript BDFL is to serve as support in the decision making process, and provide a compass to help guide the project towards its vision and its core principles. It is not meant to replace the group of maintainers in the decision process nor to discourage healthy conversation and proposals or overrule the voting process.
+
+The BDFL has the following responsibilities:
+
+-   Never use their rights in ways that discourage an open and respectful dialogue among Maintainers, Contributors, and Users.
+-   Ensure a fair and respectful voting process.
+
+and rights:
 
-**2.2. Appeal Process**. Decisions may be appealed by opening an issue and that appeal will be considered by the Maintainers in good faith, who will respond in writing within a reasonable time. If the Maintainers deny the appeal, the appeal my be brought before the Organization Steering Committee, who will also respond in writing in a reasonable time.
+-   Right to bypass the voting process towards a decision in case there isn't an unanimous consensus on a specific matter among all the other Maintainers of the project or in case they consider the decision against the project vision or best interest.
```

Comment:
Please change this to (so we have the conditional explicit at the start):

```
-  If the maintainers cannot or will not make a decision through their usual processes, the BDFL has the right to bypass such processes and be the final arbiter of the decision in hand.
```

I'm not comfortable with the bit, "in case they consider the decision against the project vision or best interest". That's way too broad and basically gives BDFL `sudo` powers by fiat. Rather, the version I've written ensures the BDFL is *always* _only_ the arbiter of last resort when there's a breakdown in the usual maintainer's process.

It's the difference between a parliamentary monarchy and absolute monarchy. In the first, the monarch is not absolute ruler, but a useful "last resort" and bureaucratic function. In the second, they can do as they please (never a good situation because there needs to be guards).

Reply:
@ntoll  As I expected, that's probably the point that requires more discussion. 

I've added the paragraph the way it is because I wanted to make sure that this BDFL model is more like a soft BDFL model where the maintainers can still overrule a BDFL decision (through a unanimous vote). That is not how usual BDFL models work. Funny enough, all the other feedback I've got so far was in the opposite direction 🤣 

In short, I've got 2 types of feedback so far (below is my interpretation... may be incorrect): 

* This sounds more like a weak and doesn't really solve the problem we have around converging toward a decision effectively. We should have the **rule** that the BDFL can bypass any vote, even unanimous, and have the ** practice ** that this rule is exercised only in very rare cases 
* This sounds like a monarchy. We don't want a BDFL but a person to resolve impasses within maintainers.

I'm open to both options (the goal here is not the proposal itself but how it can help the project achieve its vision). The only issue with your proposed text is that it's way too loose to move any discussion forward because `cannot or will not make a decision through their usual processes` doesn't define how we agree on what it means. Is this after a day? A week? What are the conditions where we say "maintainers cannot or will not make a decision through their usual processes"?

Reply:
Offered in a constructive spirit:

So, of the two bullet points, the first is definitely tyranny: one person can do whatever they like with no recourse to holding them to account. This is not a good thing and I feel deeply uncomfortable that we're even considering it. I'm not sure what I would do were we to adopt this way of working because it goes against everything I believe in. 

The second is more like what you get in the US system or the UK's parliamentary monarchy, where there's a significant person with a casting vote (the VP in the case of votes on the Hill, or the royal prerogative here in the UK). They act as the super powerful unblocker when the usual processes don't work. But the point is, their power is specific, constrained and managed. They complement (rather than replace) a system based upon collective ownership and decision making.

If there's one thing history tells us, "the practice is..." sounds well and good, until the BDFL decides otherwise, including changing all these rules. That's why the BDFL should have a very powerful but well defined and specific role in unblocking decision making (and only that). It's all about scope of influence. Just remember, in the future the "B" bit of BDFL is not guaranteed - and history (again and again and again) shows us this to be the case, including many open source projects that imploded because "I'm in charge" is not a great way to manage disputes.

FWIW my preference is for something like the current CPython situation: a steering council, elected by maintainers, containing an odd number (I suggest 3), who arbitrate and decide by simple majority when consensus cannot be reached by the maintainers. It's mature, sustainable and representative. Since it's a council, it also removes making it all about an individual personality while ensuring the heavy burden of responsibility is shared. It's also just as lightweight as having a BDFL too. As Winston Churchill said )paraphrasing from memory), "democracy is the very worst form of governance, except for all the others that have been tried".

Anyway... I have a very strong dislike of tyranny for deeply personal moral and ethical reasons and I feel very uncomfortable about it. If we must have a BDFL - their role needs to be clear, constrained and with an escape hatch so they can be removed.

Put another way, an unconstrained BDFL falls short when answering these famous five questions concerning power:

* What power have you got?
* Where did you get it from?
* In whose interest do you exercise it?
* To whom are you accountable?
* How can we get rid of you?

They can only answer in the follow way:

* Unlimited power (I'm a BDFL).
* Because I'm BDFL with unlimited power.
* My own - that's the point... I exercise my judgement for what *I see* is the good of the project, and you just have to lump it.
* Nobody.
* You can't (or you walk away / fork the project / watch it implode etc...).

I hope you agree, none of these answers are the sort we want to give to anyone wanting to join our project. Especially when the only recourse to go against the BDFL is to walk away or fork the project. Our community is littered with projects that imploded because the BDFL system alienated people rather than an alternative way of working that empowered, engaged and nurtured them.

And to those who say, "but it's only a software project, we just want someone to make decisions so I can get on with the important stuff of writing code" ~ **writing software *is* a political activity** because we're organising and exercising control of resources, processes and how things unfold _through and in a digital medium_. This is why it's so important to get our own internal "political" structures well defined: _because asking what they should be tells us what sort of project we'll become_.

I don't want to work in a tyranny... no matter how nice the tyrant may be. History tells us it never ends well for all concerned, especially if the project becomes very successful (as we hope will be the case for PyScript).

Peace. :hugs:

Reply:
I'd like to address a quick thing first (because it puts this conversation under very bad and emotionally overloaded terms). 

> So, of the two bullet points, the first is definitely tyranny

That is simply not true. Tyranny and Dictatorship are 2 different words with 2 different meanings. A dictator is someone who acquires power [without the consent of the people] (in this and most cases in sw that's not true) and is (or can be) aided by a group of counselors. A Tyrant is someone who leads in an oppressive and ruthless way, in their own interest and above the interest of their community as a whole. I hope the difference between a Benevolent Dictator vs a Tyrant is very clear to everyone (and on a personal note, I do abhor the idea that anyone close to the PyScript project could be a Tyrant if elected a BDFL). 

With that out of the way, I think we are not that far apart in how we see things. Let me start getting something out there right away: I'm not advocating for a BDFL with unlimited powers. If I wanted, it was very easy to start that right away when I created the project 😉 .  Even though it's not my personal preference, I also don't think it's a bad thing per se. There are plenty of examples of software projects that work that way and they work great. (And are actually way closer to a democracy than a dictatorship, not to mention a tyranny).

I do have to acknowledge that consensus among the maintainers, like we are running things today, is not being very effective and we've been having a hard time/moving slowly in the decision-making process. Tbh, I don't think governance models are a one size fits all thing. There are BDFL-led projects that ran well with that model (and then not so great when they turned into a steering council mode) and many great projects that work well with a steering council model.

Now, going back to a point you made: **writing software is a political activity**. I usually say "Writing software is a social activity" (because politics kind of have a smell to it 😃  and I think people > politics as a whole) but we are not far off. In fact:

* the reason I wrote `Right to bypass the voting process towards a decision in case there isn't a unanimous consensus on a specific matter among all the other Maintainers of the project` was because I did want to create a mechanism for the steering council (the maintainers in this case) to limit the BDFL power
* I personally prefer a model with a steering council led by a leader (similar to a board and a chairman of the board) that has some extra rights to allow this person to [try to] avoid deadlocks, slowdowns, bureaucracy overhead, etc..  while being the compass
* While the D-word (dictator) often implies "without consent of the people", it hasn't been the case in most cases I've seen in regard to a BDFL in software
* I'm opposed to this role to created without people's consent.
* I actually was going to propose a "soft-BDFL" model where it's really more like a Chairman of a Steering council that has similar rights to what I proposed but can be voted down by the council and can only bypass votes below a certain threshold
* I did not write that proposal here because, actually, most of the feedback I had when talking with other maintainers was that they preferred a more classic BDFL role (so I wrote what I think the majority prefers)

Now, I'll repeat my preference here just to make sure it's added to the options and called out explicitly: "soft-BDFL" model (or maybe we should call it a steering council with a Chairperson) where the Chair has similar rights to what I proposed but can be voted down by the council and can only bypass votes below a certain threshold. This is not about Tyranny (dang, I can't help but hate this term) but about a group of people putting trust in a leadership group and a leader.

(Also think this is a hard conversation over text)